### PR TITLE
Added some documentation for the NGINX proxy setup in docker swarm

### LINKED ISSUE #31: Setup NGINX as an reverse proxy for the Docker Swarm and document the work.
**Is your feature request related to a problem? Please describe.**
Currently we only have a sing static NGINX service that serves the Docker Swarm. 

**Describe the solution you'd like**
Make a Docker stack that will support a dynamic NGINX reverse proxy for the Docker Swarm

**Describe alternatives you've considered**
Apache is an alternative but it is preferred if NGINX is used instead. 

**Additional context**
See server documentation for information regarding how to access the server.

### PR DESCRIPTION
**Pull request by thomaslhansen**
Fixes #31

### CODE REVIEW

**Comment by TobiasPalludan** [comment]
@aau-giraf/sw609f19 you guys have been assigned to review this PR, and @Tricky12321 has the main responsibility that it gets done. You may assign it internally in the group as you like, but I will contact @Tricky12321 about any issues.

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Proxy
+
+For the Giraf project we only have two public IP's that both has an DNS A name to srv.giraf.cs.aau.dk. The firewall settings for those ips are that only port ```80``` and port ```443``` are allowed through and since the different parts of the system uses other ports such as port ```5000``` and ```3306``` we need an reverse proxy to pass the traffic intended for those port into our Docker Swarm. In the following the configuration of the proxy will be elaborated upon.
```
Comment (fryyd) [inline ServerAdministration/DockerSwarm/Proxy.md]: Maybe change "IP's that both has an DNS A name to" to "IPs that both have a DNS name to".
Maybe change "those ips" to "those IPs".
Maybe change "we need an reverse" to "we need a reverse".
Maybe change "for those port" to "for those ports".
Maybe change "into our Docker Swarm" to "into the Docker Swarm".
Maybe change "In the following the configuration" into "In the following, the configuration".

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+networks:
+    frontend:               # creates a new network for frontend traffic
+```
+
+The code specifies a service called proxy that uses the nginx version 1.15.x and that exposes the port 80 and 443 to the network. It has a volume attached where the nginx config folder is mapped into the container. It is attached to a network called frontend witch is used to exchange information between the proxy and the API. There will also be an network for backend communication between the API and the Database server.
+
+## nginx.conf
+
+The NGINX can be used for many different purposes and one of them is as an proxy. The following config is a standard config for a reverse proxy that will enable all sites in the ```/etc/nginx/sites-enabled/``` folder.
```
Comment (MStenkaer) [inline ServerAdministration/DockerSwarm/Proxy.md]: "one of them is as a proxy"?

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    keepalive_timeout  65;
+    include /etc/nginx/sites-enabled/*;
+    gzip  on;
+    client_max_body_size 20m;
+}
+```
+
+## API
+
+The API is file under ```/etc/nginx/sites-enabled/``` specifies where the traffic for the API is supposed to go once the NGINX server receives it. We can use the ```proxy_pass http://API``` since docker uses static DNS names inside its network all traffic for the API will be directed into the API container.
```
Comment (MStenkaer) [inline ServerAdministration/DockerSwarm/Proxy.md]: "The API is a file under ```/etc/nginx/sites-enabled/``` which specifies"??
Should the last sentence be split into two?

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        volumes:            # mounts the nginx config folder inside the container as readonly
+            - ./nginx/:/etc/nginx/
+        networks:           # uses the frontend network to pass trafic into the containers
+            - frontend
+
+networks:
+    frontend:               # creates a new network for frontend traffic
+```
+
+The code specifies a service called proxy that uses the nginx version 1.15.x and that exposes the port 80 and 443 to the network. It has a volume attached where the nginx config folder is mapped into the container. It is attached to a network called frontend witch is used to exchange information between the proxy and the API. There will also be an network for backend communication between the API and the Database server.
```
Comment (MStenkaer) [inline ServerAdministration/DockerSwarm/Proxy.md]: "called frontend which is used to exchange"??
"There will also be a network for backend"??

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
+# Proxy
+
+For the Giraf project we only have two public IP's that both has an DNS A name to srv.giraf.cs.aau.dk. The firewall settings for those ips are that only port ```80``` and port ```443``` are allowed through and since the different parts of the system uses other ports such as port ```5000``` and ```3306``` we need an reverse proxy to pass the traffic intended for those port into our Docker Swarm. In the following the configuration of the proxy will be elaborated upon.
+
+## docker-stack.yml
+
+Then using the ocker stack command for deploying to an production environment the stack file by convention has the name ```docker-stack.yml```.
```
Comment (fryyd) [inline ServerAdministration/DockerSwarm/Proxy.md]: ocker?
"to an production" to "to a production"
Maybe rephrase this sentence.

**Review by MStenkaer — CHANGES_REQUESTED** [review]
The request will be finally reviewed after the comments have been resolved.

**Review by fryyd — DISMISSED** [review]
The final review will be done when the comments have been resolved.

File: ServerAdministration/DockerSwarm/Proxy.md (hunk 3 above)
Comment (Tricky12321) [reply ServerAdministration/DockerSwarm/Proxy.md]: http://API change to API HOST

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+The API is file under ```/etc/nginx/sites-enabled/``` specifies where the traffic for the API is supposed to go once the NGINX server receives it. We can use the ```proxy_pass http://API``` since docker uses static DNS names inside its network all traffic for the API will be directed into the API container.
+
+```bash
+server {
+  listen 80;
+  listen 443;
+
+  server_name srv.giraf.cs.aau.dk;
+    location / {
```
Comment (Tricky12321) [inline ServerAdministration/DockerSwarm/Proxy.md]: Location here should be 
```suggestion
    location /api/ {
```

**Review by Tricky12321 — CHANGES_REQUESTED** [review]

**Review by MStenkaer — DISMISSED** [review]

File: ServerAdministration/DockerSwarm/Network.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# Networking Docker Swarm
+
+Networking in a Docker Swarm is fairly easy. In the docker-compose.yml file simply add the name of the networks and specify which services is connected to which network. A service can be connected to 0 or more networks and a network not connected to anything does not make any sense.
+
+The Docker Swarm has two networks to separate different services in case one of them i compromised by an intruder.
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Network.md]: is compromised

File: ServerAdministration/DockerSwarm/Network.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# Networking Docker Swarm
+
+Networking in a Docker Swarm is fairly easy. In the docker-compose.yml file simply add the name of the networks and specify which services is connected to which network. A service can be connected to 0 or more networks and a network not connected to anything does not make any sense.
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Network.md]: which services are connected to which network

File: ServerAdministration/DockerSwarm/Network.md [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+The Docker Swarm has two networks to separate different services in case one of them i compromised by an intruder.
+
+The frontend network is used to exchange information between the NGINX Proxy and the API.
+
+The backend network is used to exchange information between the API and database.
+
+Since the database has the most sensitive information that we want to protect it makes sense to isolate it from the open internet.
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Network.md]: that we want to protect, it makes sense (comma)

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
+# Proxy
+
+For the Giraf project we only have two public IPs that both have a DNS name to srv.giraf.cs.aau.dk. The firewall settings for those IPs are that only port ```80``` and port ```443``` are allowed through and since the different parts of the system uses other ports such as port ```5000``` and ```3306``` we need a reverse proxy to pass the traffic intended for those port into the Docker Swarm. In the following, the configuration of the proxy will be elaborated upon.
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: intended for those ports

Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: the system uses other ports,(comma) such as port 5000 and 3306,(comma) we need

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+## docker-compose.yml
+
+Then using the Docker Stack command for deploying to a production environment an ```.yml``` file has to be passed to the command.
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: Using the Docker Stack (remove "Then")

Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: to a production environment,(comma) a .yml file (a instead of an)

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            - ./nginx/:/etc/nginx/
+        networks:           # uses the frontend network to pass traffic into the containers
+            - frontend
+
+networks:
+    frontend:               # creates a new network for frontend traffic
+    backend:                # creates a new network for backend traffic
+```
+
+The code specifies a service called proxy that uses the nginx version 1.15.x and that exposes the port 80 and 443 to the network. It has a volume attached where the nginx config folder is mapped into the container. The networks is elaborated in the [section about network](./Network.md).
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: The networks are elaborated

File: ServerAdministration/DockerSwarm/Proxy.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    keepalive_timeout  65;
+    include /etc/nginx/sites-enabled/*;
+    gzip  on;
+    client_max_body_size 20m;
+}
+```
+
+## API
+
+The API is file under ```/etc/nginx/sites-enabled/API``` specifies where the traffic for the API is supposed to go once the NGINX server receives it. We can use the ```proxy_pass http://API/``` since docker uses static DNS names inside its network all traffic for the API will be directed into the API container. This means that the API can be accessed though http://srv.giraf.cs.aau.dk/API/ on both port ```80``` and port ```443```.
```
Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: The API file under (remove is)

Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: API can be accessed through

Comment (msimon16) [inline ServerAdministration/DockerSwarm/Proxy.md]: "We can use the ```proxy_pass http://API/``` since docker uses static DNS names inside its network all traffic for the API will be directed into the API container."
both the first part and last part of the sentence, seem to depend on "since". Rephrase.

**Review by msimon16 — CHANGES_REQUESTED** [review]

**Review by msimon16 — APPROVED** [review]

**Comment by Tricky12321** [comment]
It looks good!

**Review by Tricky12321 — APPROVED** [review]
